[Chun Khai] iP

chunkhaik opened the pull request:
# DobbyPro

>"And if Dobby does it wrong, Dobby will throw himself off the top most tower! " -Dobby the Elf, Harry Potter

DobbyPro frees your mind of having to remember things you need to do. It's

- text-based
- easy to learn
- ~~FAST~~ _SUPER_ FAST to use

All you need to do is, 
1. download it from [here](https://github.com/avock/ip/releases).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you :elf:

And it is **FREE**!

Features:

- [ ] Managing deadlines (coming soon)
- [ ] Managing tasks
- [ ] Reminders (coming soon)

If you are a Java programmer, you can use it to practice Java too! Here's the `main` method: 
````java
    public static void main(String[] args) {
      Application.launch(Main.class, args);
      System.exit(0);
    }
````

Commit:
Add Gradle support

chunkhaik committed:
Level 1: Renamed Duke as Bobby, added basic chat functionality

chunkhaik committed:
Level-2: Added add,list funcionality

chunkhaik committed:
Renamed to Dobby, Added functionality to mark/unmark as done

chunkhaik committed:
Level-4: Added new Task types (Todo, Event, Deadline) for Dobby

chunkhaik committed:
Level-5: Added Error Handling capability for Dobby (Empty task, missing date, erroneous command etc)

chunkhaik committed:
Level-6: Added delete functionality, along with relavent error handling methods

chunkhaik committed:
Level-6: Cleaned up functions and slight bug fixes

chunkhaik committed:
Level-6: Added functionality to add/mark/unmark all at once + small bug fixes

chunkhaik committed:
Level-6: Added alternative commands (quit, end, remove)

chunkhaik committed:
A-TextUiTesting: Updated automated testing functionality

chunkhaik committed:
Level-7: Add save(), Updated print methods for events

chunkhaik committed:
Level-7: Update load() function

chunkhaik committed:
Level-7: Update load() function

chunkhaik committed:
Level-8: Add Dates and Times function, Updated Load function

chunkhaik committed:
Level-8: Error when merging

chunkhaik committed:
A-MoreOOP: Add Parser class to read user input

chunkhaik committed:
A-MoreOOP: Before Deleting Old Methods

chunkhaik committed:
A-MoreOOP: Before Deleting Old Methods

chunkhaik committed:
A-MoreOOP: Update start()

chunkhaik committed:
A-MoreOOP: Update project OOP model

chunkhaik committed:
A-Packages: Add packages for project

chunkhaik committed:
small error

chunkhaik committed:
Level-9: Add find() function

chunkhaik committed:
A-CodingStandard: Update code to comply to coding standard

chunkhaik committed:
A-JavaDoc: Add javadocs to each non-private class/methods and no-trivial private methods

chunkhaik committed:
A-CodingStandard: Update codes to comply with coding standard

chunkhaik committed:
Merge remote-tracking branch 'upstream/add-gradle-support' into HEAD

chunkhaik committed:
Rename directory from Dobby to dobby

chunkhaik committed:
A-JUnit: Add JUnit tests

chunkhaik committed:
Level-7: Update ability to save to folders which have not been created

chunkhaik committed:
no message

chunkhaik committed:
Merge branch 'branch-A-CodingStandard'

# Conflicts:
#	src/main/java/dobby/Dobby.java
#	src/main/java/dobby/DobbyIO.java
#	src/main/java/dobby/commands/ByeCommand.java
#	src/main/java/dobby/commands/TaskCommand.java

chunkhaik committed:
Delete folder to be ignored: data/dobbyList.txt

chunkhaik committed:
Add data folder to .gitignore

chunkhaik committed:
Update README: changed some basic information

chunkhaik committed:
Add GFMD comment template

Evande1 commented on src/main/java/dobby/DobbyChat.java:
```diff
@@ -0,0 +1,210 @@
+     * @param list list where task is deleted from
+     */
+    public static void deleted(Task task, DobbyList list) {
+        if (list.isEmpty()) {
```
Good validation!

Evande1 commented on src/main/java/dobby/Parser.java:
```diff
@@ -0,0 +1,200 @@
+     * @param dateOutputPattern Specified output date format
+     * @return
+     */
+    public static String dateFormat(String dateString, String dateInputPattern, String dateOutputPattern) {
```
Perhaps we should use a verb like formatDate instead?

Evande1 reviewed (commented):
Generally well done, easy on the eyes and great verb usage for functions and nouns for variables. Camelcasing done well too.

changjy44 commented on src/main/java/dobby/Dobby.java:
```diff
@@ -0,0 +1,57 @@
+        DobbyChat.sayHello();
+        boolean isBye = false;
+
+        while (!isBye) {
+            String task = ui.readCommand();
+            Command cmd = Parser.parse(task);
+            cmd.execute(dobbyList, ui);
+            isBye = cmd.isBye();
+        }
```
Very simple and readable code! Easy for me to follow the happy path.

changjy44 commented on src/main/java/dobby/Dobby.java:
```diff
@@ -0,0 +1,57 @@
+     * @throws IOException
+     */
+    private static void dobbyStart() throws IOException {
+        DobbyChat.sayHello();
```
It is not immediately obvious what DobbyChat is, I didnt know that it was a static method. Maybe can write comments here?

changjy44 commented on src/main/java/dobby/DobbyChat.java:
```diff
@@ -0,0 +1,210 @@
+ * DobbyChat is a class that handles various outputs Dobby will give.
+ */
+public class DobbyChat {
+    private static final String hello = "Hello! I'm Dobby\n"
+            + "\t" + "How can I help you?";
+    private static final String bye = "Byebye. Dobby will miss you!";
+    private static final String line = "\t----------------------------------------------";
```
Maybe can put CAPS for constant variables?

changjy44 commented on src/main/java/dobby/DobbyChat.java:
```diff
@@ -0,0 +1,210 @@
+     *
+     * @param s task to be marked
+     */
+    public static void marked(String s) {
```
Naming of s can be more explicit, does s refer to the String name of the task?

changjy44 commented on src/main/java/dobby/Parser.java:
```diff
@@ -0,0 +1,200 @@
+    public static String getDate(String rest) {
+        String dateFormatted = "";
+        try {
+            int i = rest.indexOf("/");
```
Can use a better name for index of "/", "i" might be ambiguous

changjy44 commented on src/main/java/dobby/UserInput.java:
```diff
@@ -0,0 +1,90 @@
+        } else {
+            taskType = Parser.getCmd(input);
+            rest = Parser.getRest(input);
+            if (taskType.equals("mark") || taskType.equals("unmark") || taskType.equals("delete")) {
+                ind = Integer.parseInt(rest);
+            } else if (taskType.equals("todo") || taskType.equals("find")) {
+                desc = rest;
+            } else if (taskType.equals("deadline") || taskType.equals("event")) {
+                String dateType = Parser.getDateType(rest);
+                if (dateType.equals("noDate")) {
+                    date = "noDate";
+                    desc = Parser.getDesc(rest);
+                } else if (taskType.equals("deadline") && !(dateType.equals("by"))) {
+                    date = "wrongDeadline";
+                    desc = Parser.getDesc(rest);
+                } else if(taskType.equals("event") && !(dateType.equals("at"))) {
+                    date = "wrongEvent";
+                    desc = Parser.getDesc(rest);
+                } else {
+                    date = Parser.getDate(rest);
+                    desc = Parser.getDesc(rest);
+                }
```
Can be clearer and set each case in a block

changjy44 commented on src/main/java/dobby/commands/ByeCommand.java:
```diff
@@ -0,0 +1,29 @@
+import dobby.DobbyList;
+import dobby.UserInput;
+
+import dobby.*;
```
Can make the import more explicit!

changjy44 reviewed (commented):
Overall well done! I like the identity you created for your dobby, very emotional and engaging. Naming can be improved, can use more intuitive names.

chunkhaik committed:
Level-10: Add UI to Dobby

chunkhaik committed:
A-CheckStyle: Update code to comply with codestyle

chunkhaik committed:
Update: Fixed all wrong indentation

chunkhaik committed:
Fix: Small indentation errors

chunkhaik committed:
C-FriendlierSyntax: Add new simplify( ) function

Some users may find commands too long and troublesome

Previously, there were no ways to customizer user's preferred commands

Now, we add a method to customize user's preferred comands

chunkhaik committed:
Merge pull request #1 from avock/heads/C-FriendlierSyntax

Heads/c friendlier syntax

chunkhaik committed:
Update code quality

chunkhaik committed:
Add Continuous Integration (CI) support

chunkhaik committed:
Update javadocs

chunkhaik committed:
Update coding style

chunkhaik committed:
Update CI: Removed unrelated I/O redirection tests

chunkhaik committed:
Minor fix on abstraction

chunkhaik committed:
A-Assertion: Add assertion feature

Document assumptions at certain points in the code that should hold no matter what the user's input is.

Include asserts in method to ensure arguements are not null.

This will ensure subsequent methods that are linked to this are valid and do not cause further bugs

Minor javaDoc updates has been added as well

chunkhaik committed:
Fix Dobby writing of taskList to external file

Previously, Dobby only saves on /quit or /bye. However, some user's may force quit without saving, causing the task list to not be captured and saved

Update Dobby to save after every command that changes or alter the task list

chunkhaik committed:
A-Assertion: Add assertion feature

Document assumptions at certain points in the code that should hold no matter what the user's input is.

Include asserts in method to ensure arguements are not null.

This will ensure subsequent methods that are linked to this are valid and do not cause further bugs

Minor javaDoc updates has been added as well

chunkhaik committed:
Merge pull request #2 from avock/heads/A-Assertions

A-Assertion: Add assertion feature

chunkhaik committed:
Fix small checkstyle error

chunkhaik committed:
Merge branch 'master' of https://github.com/avock/ip

chunkhaik committed:
Fix small checkstyle error

chunkhaik committed:
Update GUI:

Dialog box would overflow if overlong text is entered. The overall outlook of the chat bubbles are really dull and squarish.

- add maximum width to messages
- update GUI color and chat bubble shape
- add padding between chat bubble and chat icon
- update font of messages

chunkhaik committed:
Add more personality to Dobby

- Update some replies to be more friendly
- Update the "command" feature to be more accurate
- Fix some checkstyle issues

chunkhaik committed:
Fix bug in parsing user input

chunkhaik committed:
A-UserGuide: Set up product website

- Included essential information for user to use Dobby in a user guide, as well as sample input and outputs.
- Included screenshot of Dobby's GUI.
- Updated Dobby to the latest version

chunkhaik committed:
Fix small errors:

- line space error in README
- handling case where user doesn't provide task description for event and deadline

chunkhaik committed:
Fix small errors:
- line space error in README
- handling case where user doesn't provide task description for event and deadline

chunkhaik committed:
Update GUI: Make GUI more user friendly

Previous GUI was too dull and chat bubbles of the user and chatbot were easily mixed up.

- Seperated colors of user and chatbot chat bubbles
- Added application logo and title

chunkhaik committed:
Update folder name of stored data

Previously, data was stored into folder named "data" which was unintuitive, as feedbacked by my peeR

Updated data to be stored into folder "dobbyData" instead

chunkhaik committed:
Fix User guide : small markdown syntax error

chunkhaik committed:
Fix simplify command bug, and some long methods

chunkhaik committed:
Update simplify command to work with list-command

chunkhaik committed:
Fix help button commands

chunkhaik committed:
Fix empty user input bug

chunkhaik committed:
Fix code quality issue

chunkhaik committed:
Update UI

Matters fixed:
- update image used to a non-copyrighted image (source: Unsplash)
- unresizable windows
- background color difference during initial startup
- limit window resizing to a minimum to allow to view both user and dobby's output
- update chat bubble to a more friendly color

chunkhaik committed:
Fix code quality -
